### PR TITLE
Reworking of the Users/Roles/Collections models

### PR DESCRIPTION
**Pull request by jamalex**
## Summary

Some reworking, as discussed, of the Users/Roles/Collections models and surrounding logic.

Main motivations:
- Simplifying the implementation of Roles to be a direct M2M-style table linking `Collections` with `FacilityUsers` (eliminating `HierarchyNode` and friends).
- Adding some stubs of the Morango machinery to see how it will interact with the need for segmenting data by facility.
- Some extensions of the Django MPTT classes to make them "Morango-safe" (making tree_ids into UUIDs instead of auto-incrementing integers)
- Paving the road for integration of the class-based permissions stuff (https://github.com/learningequality/kolibri/pull/36), which will be the next PR.

## TODO

- [x] Have tests been written for the new code?
- [x] Has documentation been written/updated?
- [x] New dependencies (if any) added to requirements file
- [x] Add an entry to CHANGELOG.rst
- [x] Add yourself it AUTHORS.rst if you don't appear there

## Reviewer guidance

For some of the files, it may be preferable to just look at the new version, rather than the diff, but a quick skim through the diffs may be useful first, to get a sense of where changes have been made.

Attempts were made to keep the model interfaces similar to what they were before. The main changes were:

- The `add_learner`, `add_coach`, and `add_admin` methods were moved from the subclasses to the base `Collection` model, as we in principle allow any role type for any level of the collections.
- The `add_learner_group` and similar methods were removed, as it makes for a safer API to have the parent set from the child collection.
- Constants were added for role types and collection types.
- When creating most model types, it's now necessary to pass in a `dataset` field.

## Documentation

Definitions of some of the key concepts (Users/Roles/Collections), and illustrations of the relationships between them, here:
https://docs.google.com/document/d/1X8acSDfbOu1byeUdzAjhXoNV2RTFiAmLB3oEXfTKjtg/edit#

### CODE REVIEW

File: kolibri/auth/migrations/0001_initial.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
+# -*- coding: utf-8 -*-
```
Comment (jamalex) [inline kolibri/auth/migrations/0001_initial.py:1]: I ended up needing to delete the old migrations and create a new one, as it was getting some weird issues with migrating the changed uniqueness constraints. You can skip over this file and the next few deleted files; sorry for the bloat.

File: kolibri/auth/models.py [hunk 2]
```diff
@@ -1,22 +1,34 @@
 """
```
Comment (jamalex) [inline kolibri/auth/models.py:1]: This file may be easiest to look at as a whole (the new version) rather than a diff, as per preference.

File: kolibri/auth/models.py [hunk 3]
```diff
@@ -26,120 +38,258 @@ class KolibriValidationError(KolibriError):
     pass
 
 
-class BaseUser(AbstractBaseUser):
+##############################################
+# MORANGO: These pieces will live in Morango, but dummies/stubs are temporarily included here until Morango is ready.
+##############################################
```
Comment (jamalex) [inline kolibri/auth/models.py]: Again, the stuff in this MORANGO interlude are just stubs, and no need to review them. They will be more fully featured and documented within Morango itself, later.

File: kolibri/auth/models.py [hunk 4]
```diff
@@ -26,120 +38,258 @@ class KolibriValidationError(KolibriError):
...
+    tree_id = UUIDField()
+
+    class Meta:
+        abstract = True
+
+
+##############################################
+# /END MORANGO
+##############################################
+
```
Comment (jamalex) [inline kolibri/auth/models.py:42]: Review can start again here.

File: kolibri/auth/test/test_collections_and_roles.py [hunk 5]
```diff
@@ -2,204 +2,143 @@
...
 
-        self.f = Facility.objects.create()
-        self.f.add_admin(user3)
-        self.f.add_classroom(self.cr)
+        self.lg = LearnerGroup.objects.create(dataset=self.dataset, parent=self.cr)
+        self.lg.add_learner(user1)
 
     def test_remove_learner(self):
-        self.assertEqual(Learner.objects.count(), 1)
+        self.assertEqual(Role.objects.filter(user=self.user1, kind=Role.KIND_LEARNER, collection=self.lg).count(), 1)
```
Comment (jamalex) [inline kolibri/auth/test/test_collections_and_roles.py:35]: There will be some nicer APIs for checking roles/relations once the permissions stuff is integrated. I will spec these interfaces out in the permissions tests (and perhaps in this file as well) as the first step of integrating that stuff.

File: kolibri/auth/test/test_permissions.py [hunk 6]
```diff
@@ -1,454 +1,454 @@
-from __future__ import absolute_import, print_function, unicode_literals
```
Comment (jamalex) [inline kolibri/auth/test/test_permissions.py:1]: Commented out the tests in this file for now -- I'll be writing tests from scratch, but useful to have this as a reference until that's done.

File: kolibri/auth/models.py [hunk 7]
```diff
@@ -26,120 +38,258 @@ class KolibriValidationError(KolibriError):
...
+    class Meta:
+        abstract = True
+
+    dataset = models.ForeignKey("FacilityDataset", blank=True, null=True)
+
+    def save(self, *args, **kwargs):
+        if not self.id:
+            self.id = self.calculate_uuid()
+        if self._require_dataset and not self.dataset:
+            raise IntegrityError("This model must be associated with a particular FacilityDataset.")
```
Comment (jamalex) [inline kolibri/auth/models.py]: Had a discussion with @benjaoming in the dev room about it being an anti-pattern to raise `ValidationError` in the `save` method. Since this case here is like a soft version of `null=False`, we figured a nice approach would be to raise a DB-level error like `IntegrityError`. A `ValidationError` is still raised in the `clean` method, so that it can be caught earlier and handled, if validation is taking place.

**Comment by MCGallaspy** [comment]
Would be great to incorporate documentation on the google drive into documentation in the repo. Sphinx+RTD is much more developer friendly and easily discoverable.

**Comment by benjaoming** [comment]
+ can export to PDF and EPUB!

File: kolibri/auth/models.py (hunk 3 above)
Comment (MCGallaspy) [inline kolibri/auth/models.py]: Since we're constrained by this stuff, I'm uncomfortable not giving it any scrutiny. It opens the possibility that we'll code ourselves into a hole. Must it be included at this time _at all_? If so, can we at least we can decide on an API? In other words, can we decide the interface of Morango that this module can plug in to, in order to avoid tightly coupling the implementation?

File: kolibri/auth/models.py [hunk 8]
```diff
@@ -1,145 +1,293 @@
...
+    class Meta:
+        abstract = True
+
+
+##############################################
+# /END MORANGO
+##############################################
+
+
+class KolibriSyncableModel(SyncableModel):
```
Comment (MCGallaspy) [inline kolibri/auth/models.py]: This whole class, for instance, is tightly coupled to Morango. If this is to be the base class for all or most of Kolibri's Models, should we not scrutinize it now at least to the extent of agreeing that it provides a reasonable API?

File: kolibri/auth/test/test_users.py [hunk 9]
```diff
@@ -26,33 +25,33 @@ class UserProxyManagerTestCase(TestCase):
...
         base_users = BaseUser.objects.filter(username__in=self.fu_usernames)
         facility_users = FacilityUser.objects.all()
-        self.assertListEqual(sorted([u.id for u in base_users]), sorted([u.id for u in facility_users]))
+        self.assertEqual(set([u.id for u in base_users]), set([u.id for u in facility_users]))
 
     def test_device_owner_manager_returns_device_owners(self):
-        base_users = BaseUser.objects.filter(username__in=self.do_usernames)
-        device_owners = DeviceOwner.objects.all()
-        self.assertEqual(len(base_users), len(device_owners))
+        device_owner = DeviceOwner.objects.all()[0]
```
Comment (MCGallaspy) [inline kolibri/auth/test/test_users.py]: There's a `first` method for this.

File: kolibri/auth/test/test_users.py [hunk 10]
```diff
@@ -26,33 +25,33 @@ class UserProxyManagerTestCase(TestCase):
...
+        self.assertIsInstance(device_owner, DeviceOwner)
 
     def test_do_set_correct(self):
         """ continuation of test_device_owner_manager_returns_device_owners """
         base_users = BaseUser.objects.filter(username__in=self.do_usernames)
         device_owners = DeviceOwner.objects.all()
-        self.assertListEqual(sorted([u.id for u in base_users]), sorted([u.id for u in device_owners]))
+        self.assertEqual(set([u.id for u in base_users]), set([u.id for u in device_owners]))
```
Comment (MCGallaspy) [inline kolibri/auth/test/test_users.py]: The previous way (`assertListEqual`) also ensures the length of the querysets is the same.

File: kolibri/auth/test/test_users.py [hunk 11]
```diff
@@ -60,12 +59,18 @@ class UserSanityTestCase(TestCase):
...
-
-    def test_base_user(self):
-        with self.assertRaises(NotImplementedError):
-            BaseUser().is_device_owner()
+        self.dataset = FacilityDataset.objects.create()
+        self.user = FacilityUser.objects.create(
+            username="mike",
+            first_name="Mike",
+            last_name="Gallaspy",
+            password="###",
```
Comment (MCGallaspy) [inline kolibri/auth/test/test_users.py:19]: Use [set_password](https://docs.djangoproject.com/en/1.9/ref/contrib/auth/#django.contrib.auth.models.User.set_password) unless you don't care about authentication. The password field is, IIRC, actually the hash of the raw password, so if you set it like this you won't be able to authenticate.

**Comment by MCGallaspy** [comment]
My comments are split into two parts:

Interface
---------

The interface as demonstrated in the tests. For a lot of Models (anything inheriting from `KolibriSyncableModel`) it's necessary to specify a `dataset` field. I understand the necessity of this -- it's a requirement for interfacing with Morango that can't be handled in a general way easily. But there *are* still several special cases that crop up where we *can* hide the details of `dataset` management. Requiring future devs to deal with `dataset`s unnecessarily creates an avoidable cognitive burden. We should aim to make using these Models as simple as possible -- if we can hide the details in many usual cases then we benefit by avoiding potential future errors (and a lot of typing).

I have the following suggestions to that end:

1. For `Collection`s its hard to generalize `dataset` management. But for the proxy models it's easy. There's a one-to-one correspondence between the `Facility` Model and `dataset`s -- so create a `dataset` implicitly when you create a `Facility`. For subcollections, if we require `parent` is non-null, then we can infer the dataset implicitly by finding the root `Facility`. We can hide `dataset` management by encouraging using the proxies, while still allowing flexibility by using the base `Collection`.

1. For `Role`s of all types, because they represent a many-to-many relationship with `Collection`s, we can infer the `dataset` through the `Collection`.

2. For `FacilityUser`s if we require a foreign-key to a `Facility`, then the dataset can again be inferred. Our code will read more naturally and be less coupled to Morango if we enforce having an associated `Facility` instead of requiring the `dataset` to be set. Compare the following interfaces:

```python
dataset = FacilityDataset.objects.create()
# It's an error to forget the dataset -- but that might be surprising!
# user = FacilityUser.objects.create(username="foo")
user = FacilityUser.objects.create(username="foo", dataset=dataset)

# How can I get to a user's facility?
# Through the roles? What if they don't have any roles?
user.role_set.filter(collection__kind=Collection.KIND_FACILITY)
# There's no way to get at it through the dataset either, which
# is a little counterintuitive -- FacilityDataset should be in 1-to-1
# correspondence with Facilities.
```

```python
facility = Facility.objects.create()
# If we require a non-null ForeignKey field, forgetting the facility is an error, too!
# This is de-coupled from Morango.
# Plus, it's in the name of the model.
user = FacilityUser.objects.create(username="foo", facility=facility)

# I can get a user's facility very simply!
user.facility
```

Implementation
--------------

**Collection proxy model methods vs. base Collection methods**:
I get the appeal of moving all the proxy model methods onto the base `Collection`. It makes it really easy to add *any* role type to *any* kind of `Collection`. It's nice to have that flexibility built in. But in this case I think having too much flexibility could lead to mistakes. We clearly identified what it means to be an "admin for a facility" (and coach->classroom, learner->learner group). It makes a certain sense to allow e.g. admin->classroom, even though we haven't really discussed the implications. But other cases are more ambiguous. For instance, what does it mean to have an "admin for a learner group" or a "student for a facility"? 

My suggestion is to move methods like `add_admin` and `add_coach` back onto the proxy models -- then it becomes explicit which relationships are canonical and which are ambiguous. Fully general `Role`s can still be created by using the `Collection.add_user` method, so no flexibility is lost, and if we encourage using the proxy models then we can catch simple mistakes.

**DeviceOwner and `_require_dataset` attribute**:
The only place where `_require_dataset` is set to `False` is for DeviceOwners. More specifically, it's problematic that `DeviceOwner` should inherit from `SyncableModel` at all when it's explicitly the intent that `DeviceOwners` are *never* synced. Handling this single case leads to extra complexity in `KolibriSyncableModel`. The solution is for `DeviceOwner` to *not* inherit from `KolibriSyncableModel` and remove the logic for handling that case. We had discussed having two separate non-abstract user Models -- I think this is a really compelling reason to pursue that. I was worried earlier that it might not play nicely with `django.contrib.auth`, but based on our previous discussions and the example here I think it's preferable.

**Comment by MCGallaspy** [comment]
I'm looking also at the test failures. I'll try to submit a patch so you can rebase.

**Comment by MCGallaspy** [comment]
+1 on documentation. Please commit something to the repo.

File: kolibri/auth/models.py (hunk 3 above)
Comment (jamalex) [inline kolibri/auth/models.py]: Looking through this block, I don't think anything in the rest of the code in the PR actually really depends on it. The role it played for me here was in exploring how best to decouple Kolibri and Morango (i.e. what pieces belonged on which side of the fence), as well as to think through some of the potential interactions between syncing and things like MPTT (with possible `tree_id` collisions).

It sounds like it would be cleanest if I removed this entire MORANGO block for now, and moved it to the notes/specs for Morango, so that it's clear that all of this also works without Morango (other than syncing, of course). The FacilityDataset stuff would remain, but that's not _part_ of Morango -- it serves three purposes:
- Scoping data for use by permissions. For example, until a user has a learner role for a class/group, the associated FacilityDataset is the main way we would determine who has permission to edit that user.
- Storing metadata about the facility -- since Collections only have a few key fields (like name, kind, and then the MPTT stuff), we need somewhere to store additional metadata.
- Filtering/partitioning for Morango, to make it easy/efficient to query for all model instances that are part of a specific set of "facility data". So, Morango will be told to _use_ this field, but they're not tightly coupled.

File: kolibri/auth/test/test_users.py (hunk 11 above)
Comment (jamalex) [inline kolibri/auth/test/test_users.py:19]: Yep -- the tests here don't do anything with authentication, so it shouldn't be necessary; the only reason I needed to set something for the password field was that these tests call `full_clean`, which throws a validation error if the field is empty.

File: kolibri/auth/test/test_users.py (hunk 10 above)
Comment (jamalex) [inline kolibri/auth/test/test_users.py]: Ah, interesting -- so, protecting against duplicate records, then (otherwise, set comparison also does that). Sounds good, will revert.

File: kolibri/auth/test/test_users.py (hunk 9 above)
Comment (jamalex) [inline kolibri/auth/test/test_users.py]: :+1:

File: kolibri/auth/models.py (hunk 8 above)
Comment (jamalex) [inline kolibri/auth/models.py]: The only things here that actually tie into Morango's API are `_morango_partition_fields` (which isn't actually used yet, for anything else in this PR) and a call to `calculate_uuid` (which can actually be moved to the super class' `save` method).

As mentioned above, I can update this PR to completely remove all references to Morango. It won't require many changes. This class will be renamed to `FacilityDataBaseModel`, and some pieces will be removed -- it will still have the `dataset` foreign key field, which (as noted in a comment above), while it will be _used_ by Morango, doesn't depend on Morango, and also serves other purposes.

**Comment by jamalex** [comment]
Thanks for the helpful notes!

I'm partway through addressing the comments (in particular, I addressed the inline code comments, and then noticed the high-level comments here). Will respond and address those shortly, but taking a break for now.

**Comment by jamalex** [comment]
> But there are still several special cases that crop up where we can hide the details of dataset management.

I agree that it would be nice to avoid requiring devs to specify the dataset as an argument, when creating models for which the dataset can be inferred. I was thinking about having the `clean` and `save` methods infer the dataset if possible, and set it themselves, rather than throwing an error if it's missing. Because the logic for inferring the dataset varies by model, I'm thinking to add a method called `infer_dataset`, which can get overridden to return the appropriately inferred dataset for the current model. I'll try that out.

> We clearly identified what it means to be an "admin for a facility" (and coach->classroom, learner->learner group). It makes a certain sense to allow e.g. admin->classroom, even though we haven't really discussed the implications. But other cases are more ambiguous. For instance, what does it mean to have an "admin for a learner group" or a "student for a facility"?

Yeah, we discussed the use cases for having an admin for a classroom (allowing a teacher to manage accounts and groups specifically for their class) and a coach for a facility (allowing a coach to view data for *any* student, without getting write access), but there are a few implications, yeah. I'm not worried about the backend; it will be simple and elegant to check permissions -- given how I've described it in the class-based perms spec doc (having a role for a collection means having that role for all subcollections, and means having that relation to any learners for those subcollections as well). The part I do worry about is the "user management" interfaces for administering these roles, which gets messy if it's not constrained. So that would motivate doing what you're saying, for me (constraining what's "easy" to do in the Python API). But if we are allowing a few different variations of pairings (e.g. admin for both facility and classroom), then it means defining that method multiple times, which isn't super DRY. But the code being repeated is minimal, so perhaps it's worth it. Happy to do that.

> We had discussed having two separate non-abstract user Models -- I think this is a really compelling reason to pursue that. I was worried earlier that it might not play nicely with django.contrib.auth, but based on our previous discussions and the example here I think it's preferable.

Happy to try that again -- it did feel a bit weird when I tried it last time (having `request.user` not always be an instance of `AUTH_USER_MODEL`), but given the reading we did, it shouldn't be a problem in practice, since we're providing our own auth backend. I can try a pass of that in this PR.

**Comment by MCGallaspy** [comment]
> I agree that it would be nice to avoid requiring devs to specify the dataset as an argument, when creating models for which the dataset can be inferred...

Cool, I would say it's more appropriate to implement inference in the `save` method, since `clean` must be called explicitly.

> Yeah, we discussed the use cases ... Happy to do that.

Great.

> Happy to try that again -- it did feel a bit weird when I tried it last time (having request.user not always be an instance of AUTH_USER_MODEL), but given the reading we did, it shouldn't be a problem in practice, since we're providing our own auth backend. I can try a pass of that in this PR.

Then perhaps both user types should derive from an abstract base user class while only one of them *additionally* derives from the `SyncableModel` base class? Then they could both implement "user-specific" methods, minimizing the potential difference between `request.user`s. My objection is to having `DeviceOwner` derive from `SyncableModel`. It is *not* syncable and only serves to make `SyncableModel` more complicated.

File: kolibri/auth/models.py (hunk 3 above)
Comment (MCGallaspy) [inline kolibri/auth/models.py]: > It sounds like it would be cleanest if I removed this entire MORANGO block for now

Yes, if it is not essential then remove it.

Comment (MCGallaspy) [inline kolibri/auth/models.py]: That said, regarding the justification you give for `FacilityDataset` I notice that two of the reasons have nothing to do with Morango, and one does. I caution against creating a class that has unrelated responsibilities. There _is_ a concrete justification for this class arising purely from within this module, as you point out -- what I'm suggesting is to stick close to those reasons.

For instance, does it make sense for a `FacilityDataset` to exist without a `Collection` of kind "facility"? Should `FacilityDataset` be the "privileged" model, or should `Collection`s with kind "facility"? By "privileged" I mean should it be the foreign-key for objects that are namespaced by facilities. Sadly, without a more in-depth consideration that we don't seem to have the time for, I can only appeal to my intuition. (My intuition answers "no" and "`Collection`s", FWIW.)

As far as the implementation, what you're describing sounds very much like a [one-to-one relationship](https://docs.djangoproject.com/en/1.9/topics/db/examples/one_to_one/), so you might consider if that's appropriate here.

**Comment by jamalex** [comment]
> Cool, I would say it's more appropriate to implement inference in the save method, since clean must be called explicitly.

Since `clean` gets called before `save`, and we want to catch missing data and throw a `ValidationError` within `clean`, I'm thinking to keep it DRY by putting the logic into an `infer_dataset` method, which can be used by both `clean` and `save`. I'll try doing that, and then we can discuss it based on how it looks.

> Then perhaps both user types should derive from an abstract base user class while only one of them additionally derives from the SyncableModel base class

Sounds good. But we can't have `AUTH_USER_MODEL` point to an abstract class (I believe I tried that), so we'll probably need to point it to, say, `FacilityUser`. But I like the idea of an abstract mixin class to define a common interface.

File: kolibri/auth/models.py [hunk 12]
```diff
@@ -1,497 +1,428 @@
...
 
-class KolibriValidationError(KolibriError):
-    pass
+    description = models.TextField(blank=True)
+    location = models.CharField(max_length=200, blank=True)
 
+    allow_signups = models.BooleanField(default=True)
 
-class BaseUser(AbstractBaseUser):
+
```
Comment (jamalex) [inline kolibri/auth/models.py:43]: @MCGallaspy The class below this comment is new, and handles the "dataset inference" pieces we discussed. As you can see in the test files, this means that the app programmer should very rarely need to pass in a dataset, as it will be inferred at time of model creation.

File: kolibri/auth/models.py [hunk 13]
```diff
@@ -1,497 +1,428 @@
...
-    Encapsulates both FacilityUsers and DeviceOwners, which are proxy models.
+    Draws liberally from django.contrib.auth.AbstractUser, except we exclude some fields
+    we don't care about, like email.
 
-    You should prefer to use the proxy models for this class where possible.
+    This model is an abstract model, and is inherited by both FacilityUser and DeviceOwner.
     """
+
+    class Meta:
+        abstract = True
```
Comment (jamalex) [inline kolibri/auth/models.py:107]: As discussed, the base Kolibri user is now abstract (and DeviceOwner/FacilityUser are no longer proxies).

File: kolibri/deployment/default/settings/base.py [hunk 14]
```diff
@@ -191,6 +191,6 @@
     }
 }
 
-AUTH_USER_MODEL = 'kolibriauth.BaseUser'
+AUTH_USER_MODEL = 'kolibriauth.DeviceOwner'
```
Comment (jamalex) [inline kolibri/deployment/default/settings/base.py:194]: The choice for `AUTH_USER_MODEL` was fairly arbitrary, but since I'm guessing the main place it would come into play for us is in the Django admin, and `DeviceOwner` is more likely to be granted access to the Django admin than a `FacilityUser`, I might as well put that.

**Comment by jamalex** [comment]
All the requested inline and high-level changes have been made. Should be good for final review and hopefully merging now.

One thing remaining is RST documentation. I started working on that, but it got messy as the external documentation I've written (and the documentation already in there) describes the conceptual models/hierarchies/roles in tandem with the permissions system (and the permissions system updates aren't part of this PR), so I'm going to wait and finish the documentation updates as part of the permissions PR.

**Comment by MCGallaspy** [comment]
> Since clean gets called before save

I don't think this is true. From [here](https://docs.djangoproject.com/en/1.9/ref/models/instances/#django.db.models.Model.clean):

> Note, however, that like Model.full_clean(), a model’s clean() method is not invoked when you call your model’s save() method.

**Comment by MCGallaspy** [comment]
Something's a little off with the commit history here... it's a little hard to follow. It looks like some recent commits may conflict with one another as well?

File: kolibri/auth/models.py [hunk 15]
```diff
@@ -1,497 +1,428 @@
...
+    def add_admins(self, users):
+        return [self.add_admin(user) for user in users]
 
-    def save(self, *args, **kwargs):
-        self.kind = "Learner"
-        return super(Learner, self).save(*args, **kwargs)
+    def remove_admin(self, user):
+        self.remove_user(user, Role.KIND_ADMIN)
 
+    def add_coach(self, user):
```
Comment (MCGallaspy) [inline kolibri/auth/models.py:339]: I'm a little confused that `add_coach` appears as a `Facility` method (and similarly that `add_admin` appears as a `Classroom` method). Just double checking -- this is your intent?

**Comment by MCGallaspy** [comment]
One more comment. Look at the [test coverage for models.py](https://codecov.io/github/learningequality/kolibri/kolibri/auth/models.py?ref=751f32fc0def24b62b86736adb5ca286f17393bb). Is it feasible to reach 100% coverage for this file?

**Comment by jamalex** [comment]
> > Since clean gets called before save

> I don't think this is true.

Would have been clearer as "Since clean *may* be called before save". The point was that I couldn't just put the dataset inference logic directly in either `save` or `clean`, since I don't know which one is going to be called "first" (without the other having previously been called), and each one depends on the dataset being set. So that's why I put it in a separate `ensure_dataset` method that each of them calls. This also makes it easy to override (as illustrated) to define how the dataset should be inferred for a particular model.

**Comment by jamalex** [comment]
> Something's a little off with the commit history here... it's a little hard to follow. It looks like some recent commits may conflict with one another as well?

I was working to keep the commit history clean by consolidating commits and force-pushing to my branch, but it seems GitHub keeps showing the pre-consolidated commits separately (probably in case they had comments or something). The actual recent git history is just 0337691629f9bf4f88e2f6d418d762c6c5a64112 and c287fbe8048760968150250c2854b36ed22add33.

File: kolibri/auth/models.py (hunk 15 above)
Comment (jamalex) [inline kolibri/auth/models.py:339]: Yes, that was intentional. We've discussed the use cases: we want to be able to have a coach for the entire facility (who can view data for any classes, but not change anything) and admins for specific classes (e.g. to give a teacher the ability to manage students in their class, something they can't do simply as a coach).

**Comment by jamalex** [comment]
> One more comment. Look at the test coverage for models.py. Is it feasible to reach 100% coverage for this file?

Looks feasible. If I can't make it hit a line of code, that line of code may not be so necessary. :) Will try after my meetings.

File: kolibri/auth/models.py (hunk 15 above)
Comment (MCGallaspy) [inline kolibri/auth/models.py:339]: I recall we talked about the idea. I didn't consider it settled (sorry if I didn't communicate that well in my above comment).

Somehow whether the roles coach->facility and admin->classroom exist is a matter of taste. The role coach->facility is equivalent to being a coach for _all_ classrooms in a facility, and I can't think of a permission which should distinguish admin->classroom and coach->classroom (though I'll take you at your word if you say there is). For reference, see all the permissions I _could_ think of: http://kolibri.readthedocs.org/en/latest/dev/uap/permissions.html#data-access-permissions

That's my opinion, however I find that we're now discussing ~6 lines of code to death, so I leave it to you.

**Comment by jamalex** [comment]
Should be at 100% coverage for auth.models now.

File: kolibri/auth/test/test_collections_and_roles.py [hunk 16]
```diff
@@ -115,34 +122,89 @@ class CollectionsTestCase(TestCase):
...
         classroom = Classroom.objects.create(parent=self.facility)
-        classroom.add_coach(user)
-        self.assertEqual(Role.objects.filter(user=user, kind=Role.KIND_COACH, collection=classroom).count(), 1)
+        self.assertEqual(Classroom.objects.count(), 2)
+        self.assertEqual(classroom.get_facility(), self.facility)
 
     def test_add_learner_group(self):
-        classroom = Classroom.objects.create(parent=self.facility)
+        classroom = Classroom.objects.create(name="blah", parent=self.facility)
+        classroom.full_clean()
```
Comment (MCGallaspy) [inline kolibri/auth/test/test_collections_and_roles.py:172]: When should the dev call `full_clean` for a model?

File: kolibri/auth/test/test_collections_and_roles.py [hunk 17]
```diff
@@ -115,34 +122,89 @@ class CollectionsTestCase(TestCase):
...
         classroom = Classroom.objects.create(parent=self.facility)
-        learner_group = LearnerGroup.objects.create(parent=classroom)
+        learner_group = LearnerGroup.objects.create(name="blah", parent=classroom)
+        learner_group.full_clean()
         learner_group.add_learner(user)
         self.assertEqual(Role.objects.filter(user=user, kind=Role.KIND_LEARNER, collection=learner_group).count(), 1)
 
+    def test_parentless_classroom(self):
+        classroom = Classroom(name="myclass")
+        # shouldn't be valid, because no parent was specified, and Classrooms can't be the root of the collection tree
```
Comment (MCGallaspy) [inline kolibri/auth/test/test_collections_and_roles.py:186]: I think if you make this a docstring, it actually appears in the testing output... don't quote me on that, though.

File: kolibri/auth/test/test_collections_and_roles.py (hunk 16 above)
Comment (jamalex) [inline kolibri/auth/test/test_collections_and_roles.py:172]: It gets called automatically by forms, and will be called by Morango when bringing in synced data.

A dev doesn't _need_ to call it at a particular time, but it can be called if validation is desired (beyond the basic DB integrity validation that `save` does). Ideally, any code calling `full_clean` would handle `ValidationError`s that might arise.

File: kolibri/auth/test/test_collections_and_roles.py (hunk 17 above)
Comment (jamalex) [inline kolibri/auth/test/test_collections_and_roles.py:186]: I'll triple-quote you on that!

At least with `kolibri manage test`, it doesn't seem to be printing it for me, so I guess we're OK. Maybe it only prints it when there's an error in the test, for more context.

File: kolibri/auth/test/test_collections_and_roles.py (hunk 16 above)
Comment (jamalex) [inline kolibri/auth/test/test_collections_and_roles.py:172]: (Calling it here was primarily for purposes of code coverage)

Comment (MCGallaspy) [inline kolibri/auth/test/test_collections_and_roles.py:172]: Yeah, I'm not in the habit of using it, but it seems kind of critical to proper use of these models.

Comment (jamalex) [inline kolibri/auth/test/test_collections_and_roles.py:172]: It shouldn't be critical to proper use of the models -- the same stuff gets called from `save` as well, so non-validated use of the models still behaves well.
